### PR TITLE
[automate-2389] cereal connections for project deletion

### PR DESCRIPTION
**Pull request by msorens**
### :nut_and_bolt: Description: What code changed, and why?

This sub-task adds appropriate wiring for the ingest, compliance, and node-manager services for project deletion.

Note that this PR puts the infrastructure in place for all three services. However, as only the ingest service sub-task has been merged to master, that is the only one that really works at the time of writing.

### :chains: Related Resources
N/A

### :+1: Definition of Done

Upon deleting a project from the UI or the API, the `PurgeProject` method in ingest, compliance, and node-manager services is called, as evinced via the supervisor log:
```
[5][default:/src:0]# sl | grep "starting project"
[2019-12-04T22:27:42+00:00] authz-service.default(O): time="2019-12-04T22:27:42Z" level=info msg="starting project move to graveyard for id \"p1\""
[2019-12-04T22:27:43+00:00] nodemanager-service.default(O): time="2019-12-04T22:27:43Z" level=info msg="starting project purges"
[2019-12-04T22:27:43+00:00] nodemanager-service.default(O): time="2019-12-04T22:27:43Z" level=info msg="starting project delete for id \"p1\""
[2019-12-04T22:27:44+00:00] teams-service.default(O): time="2019-12-04T22:27:44Z" level=info msg="starting project purges"
[2019-12-04T22:27:44+00:00] teams-service.default(O): time="2019-12-04T22:27:44Z" level=info msg="starting project delete project \"p1\""
[2019-12-04T22:27:45+00:00] ingest-service.default(O): time="2019-12-04T22:27:45Z" level=info msg="starting project purges"
[2019-12-04T22:27:45+00:00] ingest-service.default(O): time="2019-12-04T22:27:45Z" level=info msg="starting project delete project \"p1\""
[2019-12-04T22:27:45+00:00] authn-service.default(O): time="2019-12-04T22:27:45Z" level=info msg="starting project purges"
[2019-12-04T22:27:45+00:00] authn-service.default(O): 2019-12-04T22:27:45.823Z  info    pg/adapter.go:58        starting project delete project p1
[2019-12-04T22:27:50+00:00] compliance-service.default(O): time="2019-12-04T22:27:50Z" level=info msg="starting project purges"
[2019-12-04T22:27:50+00:00] compliance-service.default(O): time="2019-12-04T22:27:50Z" level=info msg="starting project delete for id \"p1\""
[2019-12-04T22:27:50+00:00] authz-service.default(O): time="2019-12-04T22:27:50Z" level=info msg="starting project delete from graveyard for id \"p1\""
```

### :athletic_shoe: How to Build and Test the Change
Rebuild the following components:
 - authn-service
 - authz-service
 - compliance-service
 - ingest-service
 - nodemanager-service
 - teams-service

Tip: you can generate such a list automatically with this in your .bash_profile:
```
export A2PATH="$GOPATH/src/github.com/chef/automate"
alias rebuild_what='git diff `git merge-base master HEAD` HEAD --name-only "$A2PATH/components/" | awk -F/ "{print \$2}" | sort | uniq'
```

### CODE REVIEW

File: components/authn-service/tokens/pg/adapter.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func (a *adapter) CreateLegacyTokenWithValue(ctx context.Context, value string)
 
 // PurgeProject removes a project from every token it exists in
 func (a *adapter) PurgeProject(ctx context.Context, projectID string) error {
+	a.logger.Info("starting project delete for id " + projectID)
```
Comment (msorens) [inline components/authn-service/tokens/pg/adapter.go]: Added consistent logging so we can tell what is going on with this multi-service operation. It is infrequent enough and complex enough that `info` level is warranted.

File: components/compliance-service/compliance.go [hunk 2]
```diff
@@ -177,6 +178,10 @@ func serveGrpc(ctx context.Context, db *pgdb.DB, connFactory *secureconn.Factory
 		if err != nil {
 			logrus.WithError(err).Fatal("could not register project update task executors")
 		}
+		err = project_purge.RegisterTaskExecutors(cerealProjectUpdateManager, "compliance", ingesticESClient)
+		if err != nil {
+			logrus.WithError(err).Fatal("could not register project purge task executors")
+		}
```
Comment (msorens) [inline components/compliance-service/compliance.go]: Each of the three services adds this or equivalent to receive the purge directions.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 3]
```diff
@@ -191,10 +191,10 @@ func (backend *ESClient) setDailyLatestToFalse(ctx context.Context, nodeId strin
 	return errors.Wrap(err, "setDailyLatestToFalse")
 }
 
-func (backend *ESClient) DeleteProjectTag(ctx context.Context,
-	projectTagToBeDelete string) ([]string, error) {
+func (backend *ESClient) PurgeProject(ctx context.Context, projectTagToBeDelete string) error {
```
Comment (msorens) [inline components/compliance-service/ingest/ingestic/ingestic.go]: Adjusted the method signature to satisfy the necessary interface constraints.

File: components/ingest-service/backend/elastic/elastic.go [hunk 4]
```diff
@@ -42,22 +42,38 @@ func (es *Backend) Initializing() bool {
...
+		}
+		actionJobStatus, err := es.JobStatus(ctx, esActionJobID)
+		if err != nil {
+			return err
+		}
+
+		if nodeJobStatus.Completed && actionJobStatus.Completed {
+			break
+		}
+	}
```
Comment (msorens) [inline components/ingest-service/backend/elastic/elastic.go]: The ingest service has real code, as the sub-task to process it has already been merged to master, to wit the calls to `DeleteNodeProjectTag` and `DeleteActionProjectTag`.
Unlike teams and tokens, which are done synchronously, these two sub-tasks are done asynchronously, so it needs a busy-wait loop to monitor the progress and only return when both are done.
Wondering, though, if this needs a timeout check?

File: components/nodemanager-service/pgdb/project_update.go [hunk 5]
```diff
@@ -95,10 +95,10 @@ func (projectUpdate *ProjectUpdate) updateFail(err error) {
 	projectUpdate.running = false
 }
 
-func (db *DB) DeleteProjectTag(ctx context.Context,
-	projectTagToBeDelete string) ([]string, error) {
+func (db *DB) PurgeProject(ctx context.Context, projectTagToBeDelete string) error {
+	log.Infof("starting project delete for id %q", projectTagToBeDelete)
 	// TODO implement
```
Comment (msorens) [inline components/nodemanager-service/pgdb/project_update.go]: This is sub-task 2383.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 6]
```diff
@@ -191,10 +191,10 @@ func (backend *ESClient) setDailyLatestToFalse(ctx context.Context, nodeId strin
...
+	logrus.Infof("starting project delete for id %q", projectTagToBeDelete)
 	// TODO implement
```
Comment (msorens) [inline components/compliance-service/ingest/ingestic/ingestic.go]: This is sub-task 2388.

File: components/authz-service/server/v2/project_purger_workflow/project_purger.go [hunk 7]
```diff
@@ -32,10 +32,13 @@ type ProjectPurger interface {
 
 // Map of every domain service that performs project purging.
 // Any future domain that has non-rule based project data
-// needs to be added here and RegisterTaskExecutors with the same string.
+// needs to be added here and RegisterTaskExecutor *in the domain* with the same string.
```
Comment (blakestier) [inline components/authz-service/server/v2/project_purger_workflow/project_purger.go]:
```suggestion
// needs to be added here and to the RegisterTaskExecutor *in the domain* with the same string.
```

File: components/compliance-service/ingest/ingestic/ingestic.go (hunk 3 above)
Comment (blakestier) [inline components/compliance-service/ingest/ingestic/ingestic.go]:
```suggestion
func (backend *ESClient) PurgeProject(ctx context.Context, projectTagToBeDeleted string) error {
```

Comment (blakestier) [reply components/compliance-service/ingest/ingestic/ingestic.go]: (and all the way down)

File: components/compliance-service/compliance.go [hunk 8]
```diff
@@ -177,6 +178,10 @@ func serveGrpc(ctx context.Context, db *pgdb.DB, connFactory *secureconn.Factory
 		if err != nil {
 			logrus.WithError(err).Fatal("could not register project update task executors")
 		}
+		err = project_purge.RegisterTaskExecutors(cerealProjectUpdateManager, "compliance", ingesticESClient)
```
Comment (blakestier) [inline components/compliance-service/compliance.go]: Extreme nit: we name this var `projectUpdateManager` elsewhere and I like the idea of it being the same 🤷‍♂ 🙃

File: components/ingest-service/backend/elastic/elastic.go [hunk 9]
```diff
@@ -42,22 +42,38 @@ func (es *Backend) Initializing() bool {
...
 	}
 
 	esActionJobID, err := es.DeleteActionProjectTag(ctx, projectTagToBeDeleted)
 	if err != nil {
-		return []string{}, errors.Wrap(err, "Failed to start Elasticsearch Action project tags delete")
+		return errors.Wrap(err, "Failed to start Elasticsearch Action project tags delete")
 	}
 
-	log.Debugf("Started Project delete with node job ID: %q action job ID %q", esNodeJobID, esActionJobID)
+	log.Debugf("Started Project delete with [node job ID: %q, action job ID: %q]", esNodeJobID, esActionJobID)
```
Comment (blakestier) [inline components/ingest-service/backend/elastic/elastic.go]: why do we sometimes capitalize our logging and sometimes not? 😩

File: components/ingest-service/backend/elastic/elastic.go (hunk 4 above)
Comment (blakestier) [reply components/ingest-service/backend/elastic/elastic.go]: I wonder if there's a good way to notify the user that things are in an odd state if _one_ (or more) _but not all_ tasks complete ... 🤔

File: components/compliance-service/ingest/ingestic/ingestic.go (hunk 3 above)
Comment (bcmdarroch) [reply components/compliance-service/ingest/ingestic/ingestic.go]: or `projectTagToDelete`?

File: components/ingest-service/backend/elastic/elastic.go (hunk 4 above)
Comment (bcmdarroch) [reply components/ingest-service/backend/elastic/elastic.go]: timeout check seems worthwhile

**Pull request closed without merging**